Documentation/RepoApi.md

```diff
@@ -0,0 +1,137 @@
+# Standard Repo API
+
+In order to build the .NET Core product from its constiuent repos, there needs to be a standardized set of actions
+each repo can perform. Not only will this make building the repos easier, it will make it easier to onboard
+new repos. If a new repo conforms to the standardized set of actions, source-build can easily start consuming it.
+
+This document describes the basic structure of this "Repo API", and the requirements imposed on a repo that needs
+to participate in source-build.
+
+## build.cmd/sh
+
+Each repo will have two files that implement the repo API:
+
+* build.cmd
+* build.sh
+
+The `.cmd` file will be called when running on Windows, while the `.sh` file will be called on all other operating systems.
+
+These scripts will contain the logic necessary in order to build, test, publish the repo's assets.
+
+source-build will invoke these scripts on each repo passing in:
+
+1. An action to perform
+1. A set of parameters
+
+Individual repos are free to implement more features in these build scripts than what is required by source-build. For
+example, a repo may want to provide a default action. These extra behaviors are not defined by source-build.
+
+### Actions
+
+To start, there are a number of actions source-build will need the repos to implement.  The actions are:
+
+1. `Restore` - Sync any external dependencies/tools necessary to build the repo.
+1. `Build` - Assemble the product from the current source contained in the repo.
+1. `Publish` - Put the assembled product in the specified location.
+1. `Test` - Ensure the assembled product works/acts correctly.
+
+This list of actions can be added to in the future as more scenarios are implemented.
+
+source-build formats arguments using a style that is compatible with MSBuild. The reasoning for this is that the
+majority of repos have an MSBuild based build system. Using this format will allow for arguments to easily be passed
+through to MSBuild. Similarly, both action and parameter arguments are case-insensitive, just like MSBuild default
+behavior.
+
+Actions will take the form of `-t:ActionName`, just like MSBuild targets are passed to `msbuild.exe`.
+
+### Parameters
```

Comment:
I'm an advocate for a set of "known" parameters that are supported across the entire "stack".  While it's true that some repos might have their own unique parameters, there should be a shared "contract" with known parameters that always mean the same thing.  Maybe this is implied, but when we tried doing this before, this turned out to be harder than originally anticipated.

For example, when adding/subtracting/changing a parameter which is part of the contract, the implementation starts to matter to make it reasonable.

Also, there's the matter of validation.  It'd be really helpful to a) be self documenting, and b) point the dev trying to build in a helpful direction when there's a bad parameter.

Reply:
For some more info on our last attempt, check out: https://github.com/dotnet/buildtools/blob/master/Documentation/RepoCompose.md

Reply:
>there should be a shared "contract" with known parameters that always mean the same thing. Maybe this is implied, but when we tried doing this before, this turned out to be harder than originally anticipated.

Eventually I expect we will arrive at a set of known parameters as we get further along on [Epic #145](https://github.com/dotnet/source-build/issues/145) The approach we are taking is to incrementally add to and adapt the API for specific scenarios as we stage the work. The lack of a priori specification of all possibilities is intentional at this early stage.